components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/model/UserStoreAttributeMappings.java

```diff
@@ -0,0 +1,126 @@
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Data object to hold attribute mappings for all user store types.
+ */
+public class UserStoreAttributeMappings {
+
+    private Map<String, UserStoreAttributeDO> defaultUserStoreAttrMapping;
+
+    private final Map<String, Map<String, UserStoreAttributeDO>> availableUserStoreAttrMappings = new HashMap<>();
+
+    public UserStoreAttributeMappings() {
+
+    }
+
+    /**
+     * To get default user store attribute mappings.
+     *
+     * @return Map of attribute mappings.
+     */
+    public Map<String, UserStoreAttributeDO> getDefaultUserStoreAttrMapping() {
+
+        return Collections.unmodifiableMap(defaultUserStoreAttrMapping);
+    }
+
+    /**
+     * Set default attribute mappings.
+     *
+     * @param defaultUserStoreAttrMapping Map of default attribute mappings.
+     */
+    public void setDefaultUserStoreAttrMapping(Map<String, UserStoreAttributeDO> defaultUserStoreAttrMapping) {
+
+        this.defaultUserStoreAttrMapping = defaultUserStoreAttrMapping;
+    }
+
+    /**
+     * Get attribute mappings changes for the given user store name.
+     *
+     * @param userStoreName User store name.
+     * @return Map of attribute uri and attribute mappings of the given user store name.
+     */
+    public Map<String, UserStoreAttributeDO> getUserStoreAttrMappings(String userStoreName) {
+
+        return Collections.unmodifiableMap(this.availableUserStoreAttrMappings.get(userStoreName));
+    }
+
+    /**
+     * Set all available user store attribute mappings.
+     *
+     * @param availableUserStoreAttrMappings Map of user store name and their attribute mappings changes.
+     */
+    public void setAvailableUserStoreAttrMappings(Map<String, Map<String, ChangedUserStoreAttributeDO>>
+                                                          availableUserStoreAttrMappings) {
+
+        for (String userStoreName : availableUserStoreAttrMappings.keySet()) {
+            Map<String, UserStoreAttributeDO> tempMap = getModifiedAttrMap(availableUserStoreAttrMappings
+                    .get(userStoreName));
+            this.availableUserStoreAttrMappings.put(userStoreName, tempMap);
+        }
+    }
+
+    /**
+     * To merge default attribute mappings and mappings changes of other user stores.
+     *
+     * @param changedUserStoreAttrMap Attribute mapping changes which should change default values.
+     * @return Map of attribute mappings.
+     */
+    private Map<String, UserStoreAttributeDO> getModifiedAttrMap(Map<String, ChangedUserStoreAttributeDO>
+                                                                         changedUserStoreAttrMap) {
+
+        if (defaultUserStoreAttrMapping == null) {
+            return null;
+        }
+        Map<String, UserStoreAttributeDO> clonedAttrMap = SerializationUtils
+                .clone(new HashMap<>(defaultUserStoreAttrMapping));
```

Comment:
SerializationUtils.clone has a significant performance impact. This is mentioned in the java doc as well [1]. It is better if we can use JSON Serialization With Gson[1].

[1] http://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/SerializationUtils.html
[2] https://github.com/google/gson/blob/master/UserGuide.md

Reply:
e.g. https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/f5afde187f4d51f8ef35583466ebe3e08e999758/components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java#L722

Reply:
Fixed. Now gson is used to clone